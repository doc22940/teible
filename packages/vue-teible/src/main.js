import DataTable from './DataTable.vue'
import DataColumn from './DataColumn'
export { DataTable, DataColumn }
export * from 'teible'
export default DataTable
