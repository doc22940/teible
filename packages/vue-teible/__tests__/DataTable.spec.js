import { mount, createLocalVue } from '@vue/test-utils'
import { DataTable, DataColumn } from '../src/main.js'

let localVue = null

beforeAll(() => {
  localVue = createLocalVue()
  localVue.component('DataTable', DataTable)
  localVue.component('DataColumn', DataColumn)
})

afterAll(() => {
  localVue = null
})

describe('DataTable', () => {
  const generateItems = () => [{
    id: 1,
    key: 'value z (id-1)'
  }, {
    id: 2,
    key: 'value m (id-2)'
  }, {
    id: 3,
    key: 'value a (id-3)'
  }]

  it('works with sync data', () => {
    const items = generateItems()
    const wrapper = mount(DataTable, {
      propsData: { items, perPage: 2 },
      mocks: {
        idx (id) {
          return `id-${id}`
        }
      },
      slots: {
        default: `
          <data-column field="id" label="ID" :render="idx"/>
          <data-column field="key" label="Value"/>
        `
      },
      localVue
    })

    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({ items: [] })
    return wrapper.vm.$nextTick()
      .then(() => {
        expect(wrapper.html()).toMatchSnapshot()
      })
  })

  it('works with sync data using function', () => {
    const items = generateItems()
    const wrapper = mount(DataTable, {
      propsData: {
        items () {
          return {
            items,
            total: items.length
          }
        }
      },
      mocks: {
        idx (id) {
          return `id-${id}`
        }
      },
      slots: {
        default: `
          <data-column field="id" label="ID" :render="idx"/>
          <data-column field="key" label="Value"/>
        `
      },
      localVue
    })

    expect(wrapper.vm.transformed).toEqual([])
    expect(wrapper.isVueInstance()).toBeTruthy()
    return wrapper.vm.$nextTick()
      .then(() => {
        expect(wrapper.html()).toMatchSnapshot()
      })
  })

  it('works with async data', () => {
    const items = generateItems()
    const wrapper = mount(DataTable, {
      propsData: {
        items () {
          return new Promise(resolve => {
            resolve({
              items,
              total: items.length
            })
          })
        }
      },
      mocks: {
        idx (id) {
          return `id-${id}`
        }
      },
      slots: {
        default: `
          <data-column field="id" label="ID" :render="idx"/>
          <data-column field="key" label="Value"/>
        `
      },
      localVue
    })

    expect(wrapper.isVueInstance()).toBeTruthy()
    return wrapper.vm.$nextTick()
      .then(() => {
        expect(wrapper.html()).toMatchSnapshot()
      })
  })

  it('works with no columns', () => {
    const wrapper = mount(DataTable, {
      propsData: { items: generateItems() }
    })

    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('works with empty data', () => {
    const wrapper = mount(DataTable, {
      propsData: { items: [] },
      mocks: {
        idx (id) {
          return `id-${id}`
        }
      },
      slots: {
        default: `
          <data-column field="id" label="ID" :render="idx"/>
          <data-column field="key" label="Value"/>
        `
      },
      localVue
    })

    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('works with vue-i18n', () => {
    // Mocking vue-i18n because of https://github.com/kazupon/vue-i18n/issues/501
    const items = generateItems()
    const wrapper = mount(DataTable, {
      propsData: { items, perPage: 2 },
      slots: {
        default: `
          <data-column field="id" label="ID"/>
          <data-column field="key" label="Value"/>
        `
      },
      mocks: {
        $tc (key, count) {
          if (count) {
            return `${key}-${count}`
          }

          return key
        }
      }
    })

    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('emits loaded', () => {
    const items = generateItems()
    const wrapper = mount(DataTable, {
      propsData: { items: generateItems(), perPage: 1 },
      mocks: {
        idx (id) {
          return `id-${id}`
        }
      },
      slots: {
        default: `
          <data-column field="id" label="ID" :render="idx"/>
          <data-column field="key" label="Value"/>
        `
      },
      localVue
    })

    const theItem = items[0]
    theItem.$_id = theItem.id
    theItem.id = `id-${theItem.id}`

    expect(wrapper.emitted().loaded).toEqual([[{
      items: [theItem],
      total: items.length
    }]])
  })

  it('emits update:sortBy', () => {
    const wrapper = mount(DataTable, {
      props: { items: generateItems() },
      slots: {
        default: `
          <data-column field="id" label="ID"/>
          <data-column field="key" label="Value"/>
        `
      },
      localVue
    })

    wrapper.find('th').trigger('click')
    return wrapper.vm.$nextTick()
      .then(() => {
        expect(wrapper.emitted()['update:sortBy']).toEqual([['id']])
      })
  })

  it('emits update:sortDesc', () => {
    const wrapper = mount(DataTable, {
      propsData: { items: generateItems(), sortBy: 'id', sortDesc: true },
      slots: {
        default: `
          <data-column field="id" label="ID"/>
          <data-column field="key" label="Value"/>
        `
      },
      localVue
    })

    wrapper.find('th').trigger('click')
    return wrapper.vm.$nextTick()
      .then(() => {
        expect(wrapper.emitted()['update:sortDesc']).toEqual([[false]])
      })
  })

  it('emits update:filter', () => {
    const wrapper = mount(DataTable, {
      propsData: { items: generateItems(), filter: 'z' },
      slots: {
        default: `
          <data-column field="id" label="ID"/>
          <data-column field="key" label="Value"/>
        `
      },
      localVue
    })

    const input = wrapper.find('.datatable__input')
    input.element.value = 'a'
    input.trigger('input')

    return wrapper.vm.$nextTick()
      .then(() => {
        expect(wrapper.emitted()['update:filter']).toEqual([['a']])
      })
  })

  it('allows disableFiltering', () => {
    const wrapper = mount(DataTable, {
      propsData: { items: generateItems(), filter: 'z', disableFiltering: true },
      slots: {
        default: `
          <data-column field="id" label="ID"/>
          <data-column field="key" label="Value"/>
        `
      },
      localVue
    })

    expect(wrapper.find('.datatable__input').exists()).toBeFalsy()
  })

  it('keeps original items', () => {
    const items = [{
      foo: {
        bar: {
          baz: 'yo'
        }
      },
      qux: 'LOWERCASE ME'
    }]
    const wrapper = mount(DataTable, {
      propsData: {
        items
      },
      mocks: {
        u (value) {
          return value.toUpperCase()
        },
        l (value) {
          return value.toLowerCase()
        }
      },
      slots: {
        default: `
          <data-column field="foo.bar.baz" label="dot-chaining field" :sortable="false" :render="u"/>
          <data-column field="qux" label="plain field" :sortable="false" :render="l"/>
        `
      },
      localVue
    })

    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
