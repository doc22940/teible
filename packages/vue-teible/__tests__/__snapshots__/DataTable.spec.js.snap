// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable keeps original items 1`] = `
<div class="datatable">
  <div class="datatable__wrapper">
    <div class="datatable__heading">
      <div class="datatable__filter datatable__unit"><input type="text" placeholder="Filter table data" class="datatable__input">
        <!---->
      </div>
      <div class="datatable__unit datatable__text"><span>
          Showing <span>the last entry</span> of 1 records
        </span></div>
    </div>
    <div class="datatable__screen">
      <table cellspacing="0" cellpadding="0" class="datatable__content">
        <thead class="datatable__head">
          <tr>
            <th scope="col" class="datatable__column"><span class="datatable__column-text">Dot-chaining field</span></th>
            <th scope="col" class="datatable__column datatable__column--last"><span class="datatable__column-text">Plain field</span></th>
          </tr>
        </thead>
        <tbody>
          <tr class="datatable__row datatable__row--last">
            <td class="datatable__cell datatable__cell--last-row">YO</td>
            <td class="datatable__cell datatable__cell--last-column datatable__cell--last-row">lowercase me</td>
          </tr>
        </tbody>
      </table>
    </div>
    <nav class="datatable__pagination">
      <ul class="datatable__plist">
        <li class="datatable__pitem"><a href="#" aria-label="Previous" class="datatable__plink datatable__pprev datatable__plink--disabled"><span aria-hidden="true">«</span></a></li>
        <li class="datatable__pitem"><a href="#" class="datatable__plink datatable__plink--active">1</a></li>
        <li class="datatable__pitem"><a href="#" aria-label="Next" class="datatable__plink datatable__pnext datatable__plink--disabled"><span aria-hidden="true">»</span></a></li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`DataTable works with async data 1`] = `
<div class="datatable">
  <div class="datatable__wrapper">
    <div class="datatable__heading">
      <div class="datatable__filter datatable__unit"><input type="text" placeholder="Filter table data" class="datatable__input">
        <!---->
      </div>
      <div class="datatable__unit datatable__text"><span>
          Showing <span>1 to 3</span> of 3 records
        </span></div>
    </div>
    <div class="datatable__screen">
      <table cellspacing="0" cellpadding="0" class="datatable__content">
        <thead class="datatable__head">
          <tr>
            <th scope="col" class="datatable__column datatable__column--sortable"><span class="datatable__column-text">ID</span><svg version="1.1" width="12" height="16" viewBox="0 0 12 16" aria-hidden="true" class="octicon octicon-three-bars datatable__column-icon">
                <path fill-rule="evenodd" d="M11.41 9H.59C0 9 0 8.59 0 8c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zm0-4H.59C0 5 0 4.59 0 4c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zM.59 11H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1H.59C0 13 0 12.59 0 12c0-.59 0-1 .59-1z"></path>
              </svg></th>
            <th scope="col" class="datatable__column datatable__column--sortable datatable__column--last"><span class="datatable__column-text">Value</span><svg version="1.1" width="12" height="16" viewBox="0 0 12 16" aria-hidden="true" class="octicon octicon-three-bars datatable__column-icon">
                <path fill-rule="evenodd" d="M11.41 9H.59C0 9 0 8.59 0 8c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zm0-4H.59C0 5 0 4.59 0 4c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zM.59 11H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1H.59C0 13 0 12.59 0 12c0-.59 0-1 .59-1z"></path>
              </svg></th>
          </tr>
        </thead>
        <tbody>
          <tr class="datatable__row">
            <td class="datatable__cell">id-1</td>
            <td class="datatable__cell datatable__cell--last-column">value z (id-1)</td>
          </tr>
          <tr class="datatable__row datatable__row--odd">
            <td class="datatable__cell">id-2</td>
            <td class="datatable__cell datatable__cell--last-column">value m (id-2)</td>
          </tr>
          <tr class="datatable__row datatable__row--last">
            <td class="datatable__cell datatable__cell--last-row">id-3</td>
            <td class="datatable__cell datatable__cell--last-column datatable__cell--last-row">value a (id-3)</td>
          </tr>
        </tbody>
      </table>
    </div>
    <nav class="datatable__pagination">
      <ul class="datatable__plist">
        <li class="datatable__pitem"><a href="#" aria-label="Previous" class="datatable__plink datatable__pprev datatable__plink--disabled"><span aria-hidden="true">«</span></a></li>
        <li class="datatable__pitem"><a href="#" class="datatable__plink datatable__plink--active">1</a></li>
        <li class="datatable__pitem"><a href="#" aria-label="Next" class="datatable__plink datatable__pnext datatable__plink--disabled"><span aria-hidden="true">»</span></a></li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`DataTable works with empty data 1`] = `
<div class="datatable">
  <div class="datatable__wrapper">
    <div class="datatable__heading">
      <div class="datatable__filter datatable__unit"><input type="text" placeholder="Filter table data" class="datatable__input">
        <!---->
      </div>
      <div class="datatable__unit datatable__text"><span>No records</span></div>
    </div>
    <div class="datatable__screen">
      <table cellspacing="0" cellpadding="0" class="datatable__content">
        <thead class="datatable__head">
          <tr>
            <th scope="col" class="datatable__column datatable__column--sortable"><span class="datatable__column-text">ID</span><svg version="1.1" width="12" height="16" viewBox="0 0 12 16" aria-hidden="true" class="octicon octicon-three-bars datatable__column-icon">
                <path fill-rule="evenodd" d="M11.41 9H.59C0 9 0 8.59 0 8c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zm0-4H.59C0 5 0 4.59 0 4c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zM.59 11H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1H.59C0 13 0 12.59 0 12c0-.59 0-1 .59-1z"></path>
              </svg></th>
            <th scope="col" class="datatable__column datatable__column--sortable datatable__column--last"><span class="datatable__column-text">Value</span><svg version="1.1" width="12" height="16" viewBox="0 0 12 16" aria-hidden="true" class="octicon octicon-three-bars datatable__column-icon">
                <path fill-rule="evenodd" d="M11.41 9H.59C0 9 0 8.59 0 8c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zm0-4H.59C0 5 0 4.59 0 4c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zM.59 11H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1H.59C0 13 0 12.59 0 12c0-.59 0-1 .59-1z"></path>
              </svg></th>
          </tr>
        </thead>
        <tbody></tbody>
      </table>
    </div>
    <nav class="datatable__pagination">
      <ul class="datatable__plist">
        <li class="datatable__pitem"><a href="#" aria-label="Previous" class="datatable__plink datatable__pprev datatable__plink--disabled"><span aria-hidden="true">«</span></a></li>
        <li class="datatable__pitem"><a href="#" aria-label="Next" class="datatable__plink datatable__pnext datatable__plink--disabled"><span aria-hidden="true">»</span></a></li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`DataTable works with no columns 1`] = `
<div class="datatable">
  <div class="datatable__wrapper">
    <div class="datatable__heading">
      <div class="datatable__filter datatable__unit"><input type="text" placeholder="Filter table data" class="datatable__input">
        <!---->
      </div>
      <div class="datatable__unit datatable__text"><span>No records</span></div>
    </div>
    <div class="datatable__screen">
      <table cellspacing="0" cellpadding="0" class="datatable__content">
        <thead class="datatable__head">
          <!---->
        </thead>
        <tbody></tbody>
      </table>
    </div>
    <nav class="datatable__pagination">
      <ul class="datatable__plist">
        <li class="datatable__pitem"><a href="#" aria-label="Previous" class="datatable__plink datatable__pprev datatable__plink--disabled"><span aria-hidden="true">«</span></a></li>
        <li class="datatable__pitem"><a href="#" aria-label="Next" class="datatable__plink datatable__pnext datatable__plink--disabled"><span aria-hidden="true">»</span></a></li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`DataTable works with sync data 1`] = `
<div class="datatable">
  <div class="datatable__wrapper">
    <div class="datatable__heading">
      <div class="datatable__filter datatable__unit"><input type="text" placeholder="Filter table data" class="datatable__input">
        <!---->
      </div>
      <div class="datatable__unit datatable__text"><span>
          Showing <span>1 to 2</span> of 3 records
        </span></div>
    </div>
    <div class="datatable__screen">
      <table cellspacing="0" cellpadding="0" class="datatable__content">
        <thead class="datatable__head">
          <tr>
            <th scope="col" class="datatable__column datatable__column--sortable"><span class="datatable__column-text">ID</span><svg version="1.1" width="12" height="16" viewBox="0 0 12 16" aria-hidden="true" class="octicon octicon-three-bars datatable__column-icon">
                <path fill-rule="evenodd" d="M11.41 9H.59C0 9 0 8.59 0 8c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zm0-4H.59C0 5 0 4.59 0 4c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zM.59 11H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1H.59C0 13 0 12.59 0 12c0-.59 0-1 .59-1z"></path>
              </svg></th>
            <th scope="col" class="datatable__column datatable__column--sortable datatable__column--last"><span class="datatable__column-text">Value</span><svg version="1.1" width="12" height="16" viewBox="0 0 12 16" aria-hidden="true" class="octicon octicon-three-bars datatable__column-icon">
                <path fill-rule="evenodd" d="M11.41 9H.59C0 9 0 8.59 0 8c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zm0-4H.59C0 5 0 4.59 0 4c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zM.59 11H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1H.59C0 13 0 12.59 0 12c0-.59 0-1 .59-1z"></path>
              </svg></th>
          </tr>
        </thead>
        <tbody>
          <tr class="datatable__row">
            <td class="datatable__cell">id-1</td>
            <td class="datatable__cell datatable__cell--last-column">value z (id-1)</td>
          </tr>
          <tr class="datatable__row datatable__row--odd datatable__row--last">
            <td class="datatable__cell datatable__cell--last-row">id-2</td>
            <td class="datatable__cell datatable__cell--last-column datatable__cell--last-row">value m (id-2)</td>
          </tr>
        </tbody>
      </table>
    </div>
    <nav class="datatable__pagination">
      <ul class="datatable__plist">
        <li class="datatable__pitem"><a href="#" aria-label="Previous" class="datatable__plink datatable__pprev datatable__plink--disabled"><span aria-hidden="true">«</span></a></li>
        <li class="datatable__pitem"><a href="#" class="datatable__plink datatable__plink--active">1</a></li>
        <li class="datatable__pitem"><a href="#" class="datatable__plink">2</a></li>
        <li class="datatable__pitem"><a href="#" aria-label="Next" class="datatable__plink datatable__pnext"><span aria-hidden="true">»</span></a></li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`DataTable works with sync data using function 1`] = `
<div class="datatable">
  <div class="datatable__wrapper">
    <div class="datatable__heading">
      <div class="datatable__filter datatable__unit"><input type="text" placeholder="Filter table data" class="datatable__input">
        <!---->
      </div>
      <div class="datatable__unit datatable__text"><span>
          Showing <span>1 to 3</span> of 3 records
        </span></div>
    </div>
    <div class="datatable__screen">
      <table cellspacing="0" cellpadding="0" class="datatable__content">
        <thead class="datatable__head">
          <tr>
            <th scope="col" class="datatable__column datatable__column--sortable"><span class="datatable__column-text">ID</span><svg version="1.1" width="12" height="16" viewBox="0 0 12 16" aria-hidden="true" class="octicon octicon-three-bars datatable__column-icon">
                <path fill-rule="evenodd" d="M11.41 9H.59C0 9 0 8.59 0 8c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zm0-4H.59C0 5 0 4.59 0 4c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zM.59 11H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1H.59C0 13 0 12.59 0 12c0-.59 0-1 .59-1z"></path>
              </svg></th>
            <th scope="col" class="datatable__column datatable__column--sortable datatable__column--last"><span class="datatable__column-text">Value</span><svg version="1.1" width="12" height="16" viewBox="0 0 12 16" aria-hidden="true" class="octicon octicon-three-bars datatable__column-icon">
                <path fill-rule="evenodd" d="M11.41 9H.59C0 9 0 8.59 0 8c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zm0-4H.59C0 5 0 4.59 0 4c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zM.59 11H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1H.59C0 13 0 12.59 0 12c0-.59 0-1 .59-1z"></path>
              </svg></th>
          </tr>
        </thead>
        <tbody>
          <tr class="datatable__row">
            <td class="datatable__cell">id-1</td>
            <td class="datatable__cell datatable__cell--last-column">value z (id-1)</td>
          </tr>
          <tr class="datatable__row datatable__row--odd">
            <td class="datatable__cell">id-2</td>
            <td class="datatable__cell datatable__cell--last-column">value m (id-2)</td>
          </tr>
          <tr class="datatable__row datatable__row--last">
            <td class="datatable__cell datatable__cell--last-row">id-3</td>
            <td class="datatable__cell datatable__cell--last-column datatable__cell--last-row">value a (id-3)</td>
          </tr>
        </tbody>
      </table>
    </div>
    <nav class="datatable__pagination">
      <ul class="datatable__plist">
        <li class="datatable__pitem"><a href="#" aria-label="Previous" class="datatable__plink datatable__pprev datatable__plink--disabled"><span aria-hidden="true">«</span></a></li>
        <li class="datatable__pitem"><a href="#" class="datatable__plink datatable__plink--active">1</a></li>
        <li class="datatable__pitem"><a href="#" aria-label="Next" class="datatable__plink datatable__pnext datatable__plink--disabled"><span aria-hidden="true">»</span></a></li>
      </ul>
    </nav>
  </div>
</div>
`;
